codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/PaginatorGenerator.kt

```diff
@@ -272,18 +272,21 @@ private fun getItemDescriptorOrNull(paginationInfo: PaginationInfo, ctx: Codegen
     val itemMember = ctx.model.expectShape(itemMemberId)
     val isSparse = itemMember.isSparse
     val (collectionLiteral, targetMember) = when (itemMember) {
-        is MapShape ->
-            ctx.symbolProvider.toSymbol(itemMember)
-                .expectProperty(SymbolProperty.ENTRY_EXPRESSION) as String to itemMember
-        is CollectionShape ->
-            ctx.symbolProvider.toSymbol(ctx.model.expectShape(itemMember.member.target)).name to ctx.model.expectShape(
-                itemMember.member.target,
-            )
+        is MapShape -> {
+            val literal = ctx.symbolProvider.toSymbol(itemMember)
+                .expectProperty(SymbolProperty.ENTRY_EXPRESSION) as String + if (isSparse) "?" else ""
+            literal to itemMember
+        }
```

Comment:
**Question**: Don't we also need to check if the symbol is nullable alongside the map being sparse?

Reply:
No, the nullability of documents in this case is handled by the `KotlinSymbolProvider`
https://github.com/smithy-lang/smithy-kotlin/blob/main/codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/core/KotlinSymbolProvider.kt#L168-L179

Reply:
Can you clarify a little bit? I'm confused why MapShape only checks `isSparse` while CollectionShape checks `symbol.isNullable || isSparse`

Reply:
Yeah, for `MapShape` the symbol provider provides the symbol via `toSymbol` and then we use the `ENTRY_EXPRESSION` system property from the symbol as the literal, and never look at the type that the map targets like in collections. 

The `ENTRY_EXPRESSION` system property is set in the code I linked above. It uses `val valueType` which uses `val valueSuffix` to add the `?` is the value is nullable.

Reply:
I don't think that's what we want. Your new unit test `testRenderPaginatorWithSparseDocumentMap` shows the problem: It takes a sparse map shape with `Document` values and codegens a paginator that returns `Flow<Map.Entry<String, Document?>?>`.

Note that there are _two_ nullability modifiers:
* one on `Document` because `Document` is nullable, which is correct 
* one on `Map.Entry<String, Document?>` because the map is sparse, which is _incorrect_

A sparse map does not have missing _entries_, only missing _values_. I believe the logic needs to coalesce the `Document` nullability and the map sparseness into the same spot. In other words, I believe the `testRenderPaginatorWithSparseDocumentMap` test should be expecting a `Flow<Map.Entry<String, Document?>>`.

Reply:
I noticed that too, we've been code-generating them like that so I wasn't 100% sure if it was wrong. The fix is in the PR now.